File: 2026POTD/JANUARY/07-01-2026/1339.php
class Solution {
    private $total = 0;
    private $best = 0;
    private $MOD = 1000000007;

    private function sum($node) {
        if ($node === null) return 0;
        return $node->val + $this->sum($node->left) + $this->sum($node->right);
    }

    private function dfs($node) {
        if ($node === null) return 0;
        $s = $node->val + $this->dfs($node->left) + $this->dfs($node->right);
        $prod = $s * ($this->total - $s);
        if ($prod > $this->best) $this->best = $prod;
        return $s;
    }

    function maxProduct($root) {
        $this->total = $this->sum($root);
        $this->best = 0;
        $this->dfs($root);
        return $this->best % $this->MOD;
    }
}

File: 2026POTD/JANUARY/07-01-2026/1339.rs
use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn max_product(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        const MOD: i64 = 1_000_000_007;
        let mut total: i64 = 0;
        let mut best: i64 = 0;

        fn sum(node: &Option<Rc<RefCell<TreeNode>>>) -> i64 {
            if let Some(n) = node {
                let n = n.borrow();
                n.val as i64 + sum(&n.left) + sum(&n.right)
            } else {
                0
            }
        }

        fn dfs(node: &Option<Rc<RefCell<TreeNode>>>, total: i64, best: &mut i64) -> i64 {
            if let Some(n) = node {
                let n = n.borrow();
                let s = n.val as i64 + dfs(&n.left, total, best) + dfs(&n.right, total, best);
                let prod = s * (total - s);
                if prod > *best {
                    *best = prod;
                }
                s
            } else {
                0
            }
        }

        total = sum(&root);
        dfs(&root, total, &mut best);
        (best % MOD) as i32
    }
}